src/nucleus/equal.ml

```diff
@@ -104,26 +120,66 @@ let rec equal ctx ({Tt.loc=loc1;_} as e1) ({Tt.loc=loc2;_} as e2) t =
 
 and equal_ty ctx (Tt.Ty t1) (Tt.Ty t2) = equal ctx t1 t2 Tt.typ
 
-(** this function assumes that the derived signatures have already been checked equal label to label *)
-and equal_structure ~loc ctx s_def es1 es2 =
-  let rec fold ctx vs hyps fields es1 es2 =
-    match fields, es1, es2 with
-    | [], [], [] ->
-       Opt.return ctx
-
-    | (l,x,t)::fields, e1::es1, e2::es2 ->
+(** Compare signatures structurally *)
+let equal_signature ~loc ctx (s1,shares1) (s2,shares2) =
+  if Name.eq_ident s1 s2
+  then
+    Monad.lift (Value.lookup_signature ~loc s1) >!= fun s_def ->
+    (* [yts] abstracting data (types are instantiated using constraints from shares1)
+       [hyps] prove that the previous constraints were respectively equal
+          (and therefore that a type instantiated with shares1 is equal to the same instantiated with shares2)
+       [vs]  atoms of [ys1] and instantiated terms from [shares1] used to instantiate types
+       [ys1] atoms with type instantiated using shares1
+       [ys2] atoms with type instantiated using shares2
+    *)
+    let rec fold ctx hyps yts vs ys1 ys2 = function
+      | [] ->
+        context_multiabstract ~loc ctx yts >!= fun ctx ->
+        Opt.return ctx
+      | ((_,_,t),(x,None),(_,None))::rem ->
         let t = Tt.instantiate_ty vs t in
-        let e1 = Tt.instantiate vs e1 in
-        let e2 = Tt.instantiate vs e2 in
+        let jt = Jdg.mk_ty ctx t in
+        Opt.add_abstracting ~loc x jt (fun ctx y ->
+        let y1 = Tt.mk_atom ~loc y in
+        let y2 = Tt.mention_atoms hyps y1 in
+        fold ctx hyps ((y,t)::yts) (y1::vs) (y1::ys1) (y2::ys2) rem)
+      | ((_,_,t),(_,Some e1),(_,Some e2))::rem ->
+        let t = Tt.instantiate_ty vs t
+        and e1 = Tt.instantiate ys1 e1
+        and e2 = Tt.instantiate ys2 e2 in
+        (* TODO: this could be slightly more granular by using only the hyps relevant to the assumptions of t *)
         let e2 = Tt.mention_atoms hyps e2 in
-        Opt.locally (equal ctx e1 e2 t) >?= fun (ctx, hyps') ->
+        Opt.locally (equal ctx e1 e2 t) >?= fun (ctx,hyps') ->
         let hyps = AtomSet.union hyps hyps' in
-        fold ctx (e1::vs) hyps fields es1 es2
-
-    | _, _, _ ->
-       Error.impossible ~loc "equal_structure: unequal lengths"
+        fold ctx hyps yts (e1::vs) ys1 ys2 rem
+      | (_,(_,None),(_,Some _))::_ | (_,(_,Some _),(_,None))::_ -> Opt.fail
     in
-  fold ctx [] AtomSet.empty s_def es1 es2
+    fold ctx AtomSet.empty [] [] [] [] (list_combine3 s_def shares1 shares2)
+  else
+    Opt.fail
```

Comment:
It's generally easiert to understand code when the short case of an `if` statement comes first.